engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java

```diff
@@ -64,19 +75,68 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
               value.getProcessInstanceKey(), targetProcessDefinitionKey));
     }
 
+    // avoid stackoverflow using a queue to iterate over the descendants instead of recursion
+    final var elementInstances = new ArrayDeque<>(List.of(processInstance));
+    while (!elementInstances.isEmpty()) {
```

Comment:
🔧 Would be good if we create a separate method for this one. That will mitigate the complexity of the `process` method.